File: lab-experiments/cell-sedimentation/archive/rk45.ts
// Adaptive RK45 ODE solver for systems of equations
// --------------------------------------------------

export type ODEFunc = (t: number, y: number[]) => number[];

export interface RK45Result {
  t: number[];
  y: number[][];
}

export interface RK45Options {
  dt?: number;         // initial step size
  atol?: number;       // absolute tolerance
  rtol?: number;       // relative tolerance
  dtMin?: number;      // minimum step size
  dtMax?: number;      // maximum step size
  safety?: number;     // safety factor for adaptive step
  verbose?: boolean;  // enable verbose logging
  onUnderflow? : () => void; // callback on step size underflow
}

/**
 * Integrate dy/dt = f(t, y) from t0 to tEnd using adaptive RK45 (Dormand-Prince).
 * @param f rhs for dyi/dt = f(t, y)
 * @param y0 initial values for y
 * @param t0 initial time
 * @param tEnd final time to end integration
 * @param opts additional options for solver
 * @returns object containing solution { t, y } where y is an array containing each yi array to be matched with t array
 */
export function rk45_dormand_prince(
  f: ODEFunc,
  y0: number[],
  t0: number,
  tEnd: number,
  opts: RK45Options = {}
): RK45Result {
  const {
    dt = (tEnd - t0) / 1000,
    atol = 1e-6,
    rtol = 1e-3,
    dtMin = 1e-8,
    dtMax = Math.abs(tEnd - t0) / 5,
    safety = 0.9,
    verbose = false,
  } = opts;

  // Dormand-Prince coefficients
  // c_i values (nodes)
  const c2 = 1/5, c3 = 3/10, c4 = 4/5, c5 = 8/9, c6 = 1, c7 = 1;

  // a_ij values (Runge-Kutta matrix)
  const a21 = 1/5;
  const a31 = 3/40, a32 = 9/40;
  const a41 = 44/45, a42 = -56/15, a43 = 32/9;
  const a51 = 19372/6561, a52 = -25360/2187, a53 = 64448/6561, a54 = -212/729;
  const a61 = 9017/3168, a62 = -355/33, a63 = 46732/5247, a64 = 49/176, a65 = -5103/18656;
  const a71 = 35/384, a72 = 0, a73 = 500/1113, a74 = 125/192, a75 = -2187/6784, a76 = 11/84;

  // b_i values (5th order solution weights)
  const b1 = 35/384, b3 = 500/1113, b4 = 125/192, b5 = -2187/6784, b6 = 11/84, b7 = 0;
  
  // b*_i values (4th order solution weights for error estimation)
  const bs1 = 5179/57600, bs3 = 7571/16695, bs4 = 393/640, bs5 = -92097/339200, bs6 = 187/2100, bs7 = 1/40;

  let t = t0;
  let y = y0.slice();
  let h = dt;
  let err = 1;

  const tVals = [t];
  const yVals =  [y.slice()];
  // FSAL support: store k1 for reuse between attempts/steps. When a step is
  // accepted we will set `last_k1` to the computed k7 so it can be reused as
  // the k1 of the next step.
  let last_k1: number[] | null = null;

  while (t < tEnd) {
    // if (t > .13517) {
    //   let debug = 1;
    // }
    if (h !== h) {
      throw new Error("NaN encountered in h");
    }
    if (t + h > tEnd) h = tEnd - t;

    try {
      // Compute Runge-Kutta stages (FSAL): reuse last_k1 when available,
      // otherwise compute k1 and keep it so retries don't recompute it.
      let k1: number[];
      if (last_k1) {
        k1 = last_k1;
      } else {
        k1 = f(t, y.slice());
        last_k1 = k1;
      }
      const k2 = f(t + c2 * h, y.map((yi, i) => yi + h * a21 * k1[i]));
      const k3 = f(t + c3 * h, y.map((yi, i) => yi + h * (a31 * k1[i] + a32 * k2[i])));
      const k4 = f(t + c4 * h, y.map((yi, i) => yi + h * (a41 * k1[i] + a42 * k2[i] + a43 * k3[i])));
      const k5 = f(t + c5 * h, y.map((yi, i) => yi + h * (a51 * k1[i] + a52 * k2[i] + a53 * k3[i] + a54 * k4[i])));
      const k6 = f(t + c6 * h, y.map((yi, i) => yi + h * (a61 * k1[i] + a62 * k2[i] + a63 * k3[i] + a64 * k4[i] + a65 * k5[i])));
      const k7 = f(t + c7 * h, y.map((yi, i) => yi + h * (a71 * k1[i] + a72 * k2[i] + a73 * k3[i] + a74 * k4[i] + a75 * k5[i] + a76 * k6[i])));
    
      // 5th order solution
      const y5 = y.map((yi, i) => yi + h * (b1 * k1[i] + b3 * k3[i] + b4 * k4[i] + b5 * k5[i] + b6 * k6[i] + b7 * k7[i]));
      
      // 4th order solution for error estimation
      const y4 = y.map((yi, i) => yi + h * (bs1 * k1[i] + bs3 * k3[i] + bs4 * k4[i] + bs5 * k5[i] + bs6 * k6[i] + bs7 * k7[i]));

      // Error estimate and tolerance check
      const sqe = y.map((_, i) => Math.pow((y5[i] - y4[i]) / (atol + rtol * Math.max(Math.abs(y[i]), Math.abs(y5[i]))), 2));
      const ssqe = sqe.reduce((a, b) => a + b, 0);
      err = Math.sqrt(
        ssqe / y.length
      );

      // Accept step if error small enough
      if (err <= 1.0) {
        t += h;
        if (verbose) console.log(`[RK45] Accepted step to t=${t} with h=${h} (err=${err})`);
        y = y5;
        tVals.push(t);
        yVals.push(y.slice());
        // FSAL: reuse k7 as k1 for the next step (k7 is evaluated at t+h,
        // y5 for Dormand-Prince, so it can serve as the next first stage).
        last_k1 = k7.slice();
      }

      // Adapt timestep (using 5th order method)
      const scale = safety * Math.pow(1.0 / Math.max(err, 1e-10), 0.2); // 1/(p+1) where p=4
      if (scale !== scale) {
        throw new Error("NaN encountered in h adaptation");
      }
      h = Math.min(dtMax, h * Math.min(5, Math.max(0.2, scale)));
    } 
    catch (e) {
      if (verbose) {
        console.error("Error computing RK45 stages at t =", t, "with h =", h, ":", e);
      }
      err = 2;
      h = h * 0.5; // Reduce step size on error
    }

    if (h < dtMin && err > 1.0) {
      if (opts.onUnderflow) {
        opts.onUnderflow();
      }
      else {
        throw new Error("Step size underflow — integration failed");
      }
    }
  }

  return { t: tVals, y: yVals };
}

export default rk45_dormand_prince;

File: demos/HeatJacketedPFR/wasm/rk45.cpp
#include "rk45.h"

void swapPointers(double*& a, double*& b) {
    double* temp = a;
    a = b;
    b = temp;
}

RK45Result RK45Solver::solve(
        ODEFunc f,
        const double* y0,
        const size_t n,
        double t0,
        double tEnd,
        double dt,
        double dtMin,
        double dtMax,
        double atol,
        double rtol,
        double safety
) {
    if (dt < 0) dt = (tEnd - t0) / 1000.0;
    if (dtMax < 0) dtMax = std::abs(tEnd - t0) / 5.0;

    // Initialize
    double t = t0;
    double* y = new double[n];
    double* ys = new double[n];
    for (size_t i = 0; i < n; ++i) {
        y[i] = y0[i];
    }

    // First step
    double h = dt;
    double err = 1.0;
    
    // Pre allocate stage vectors and solution buffers to reuse across iterations
    double* K[7];
    for (int i = 0; i < 7; ++i) {
        K[i] = new double[n];
    }
    double* sol5 = new double[n];
    double* sol4 = new double[n];
    // For FSAL optimization
    bool has_last_k1 = false;

    // Stage pointers for readability
    double* k1 = K[0];
    double* k3 = K[2];
    double* k4 = K[3];
    double* k5 = K[4];
    double* k6 = K[5];
    double* k7 = K[6];

    while (t < tEnd) {
        // if (h != h) throw std::runtime_error("NaN in h");
        if (t + h > tEnd) h = tEnd - t;

        try {
            // Compute first stage with FSAL
            if (!has_last_k1) {
                f(t, y, k1);
                has_last_k1 = true;
            }

            // Compute all other stages
            for (int stage=1; stage<7; ++stage) {
                // Compute y for stage
                for (size_t i = 0; i < n; ++i) {
                    double a_dot_k = 0;
                    for (int j=0; j < stage; ++j) {
                        a_dot_k += A[stage][j] * K[j][i];
                    }
                    ys[i] = y[i] + h * a_dot_k;
                }
                f(t + C[stage] * h, ys, K[stage]);
                // if (VERBOSE && t == t0) std::cout << "[RK45.CPP] K[" << stage+1 << "][0]=" << K[stage][0] << std::endl;
            }

            // 5th and 4th order solutions
            for (size_t i = 0; i < n; ++i) {
                sol5[i] = y[i] + h * (b1  * k1[i] + b3  * k3[i] + b4  * k4[i] + b5  * k5[i] + b6  * k6[i] + b7  * k7[i]);
                sol4[i] = y[i] + h * (bs1 * k1[i] + bs3 * k3[i] + bs4 * k4[i] + bs5 * k5[i] + bs6 * k6[i] + bs7 * k7[i]);
            }
            // if (VERBOSE) std::cout << "[RK45] Step at t=" << t << " with h=" << h << " sol5=" << sol5[0] << " sol4=" << sol4[0] << std::endl;  

            // Error estimate
            double ssqe = 0.0;
            for (size_t i = 0; i < n; ++i) {
                double denom = atol + rtol * std::max(std::abs(y[i]), std::abs(sol5[i]));
                double sqe = std::pow((sol5[i] - sol4[i]) / denom, 2);
                ssqe += sqe;
            }
            err = std::sqrt(ssqe / n);

            // Accept step
            if (err <= 1.0) {
                t += h;
                // if (VERBOSE) std::cout << "[RK45] Accepted step to t=" << t << " with h=" << h << " (err=" << err << ")\n";
                // if (VERBOSE) std::cout << "    y[0] = " << sol5[0] << std::endl;
                // reuse sol5 as the new y by swapping to avoid allocations/copies
                // swapPointers(y, sol5);
                // Prepare k1 for next step (FSAL)
                for (size_t i = 0; i < n; ++i) {
                    y[i] = sol5[i];
                    k1[i] = k7[i];
                }
                has_last_k1 = true;
            }
            // else if (VERBOSE) {
            //     std::cout << "[RK45] Rejected step at t=" << t << " with h=" << h << " (err=" << err << ")" << std::endl;
            // }

            // Adapt timestep
            double scale = safety * std::pow(1.0 / std::max(err, 1e-10), 0.2);
            // if (scale != scale) throw std::runtime_error("NaN in h adaptation");
            h = std::min(dtMax, h * std::min(5.0, std::max(0.2, scale)));

        } catch (const std::exception& e) {
            // if (VERBOSE) std::cout << "Error at t=" << t << " with h=" << h << ": " << e.what() << std::endl;
            err = 2.0;
            h = h * 0.5;
        }

        if (h < dtMin && err > 1.0) {
            return RK45Result{ 0, t, nullptr };
            // throw std::runtime_error("Step size underflow — integration failed");
        }
    }

    // Clean up
    for (int i = 0; i < 7; ++i) {
        delete[] K[i];
    }
    delete[] sol5;
    delete[] sol4;
    delete[] ys;

    RK45Result result = {
        .length = n,
        .t = t,
        .y = y
    };
    // if (VERBOSE) std::cout << "[RK45] Integration completed at t=" << t << std::endl;

    return result;
}
